src/System.CommandLine/Parsing/CommandResult.cs

```diff
@@ -36,32 +36,147 @@ public sealed class CommandResult : SymbolResult
+                ArgumentResult argumentResult;
+
+                if (!SymbolResultTree.TryGetValue(option, out SymbolResult? symbolResult))
+                {
+                    if (option.IsRequired)
+                    {
+                        AddError(LocalizationResources.RequiredOptionWasNotProvided(option));
+                        continue;
+                    }
+                    else if (option.Argument.HasDefaultValue)
+                    {
+                        optionResult = new(option, SymbolResultTree, null, this);
+                        SymbolResultTree.Add(optionResult.Option, optionResult);
 
-                if (Command.HasArguments)
+                        argumentResult = new(optionResult.Option.Argument, SymbolResultTree, optionResult);
+                        SymbolResultTree.Add(optionResult.Option.Argument, argumentResult);
+                    }
+                    else
+                    {
+                        continue;
+                    }
+                }
+                else
                 {
-                    var arguments = Command.Arguments;
+                    optionResult = (OptionResult)symbolResult;
+                    argumentResult = (ArgumentResult)SymbolResultTree[option.Argument];
+                }
 
-                    for (var i = 0; i < arguments.Count; i++)
+                // When_there_is_an_arity_error_then_further_errors_are_not_reported
+                if (!ArgumentArity.Validate(argumentResult, out var error))
+                {
+                    optionResult.AddError(error.ErrorMessage!);
+                    continue;
+                }
+
+                if (optionResult.Option.HasValidators)
+                {
+                    int errorsBefore = SymbolResultTree.ErrorCount;
+
+                    for (var j = 0; j < optionResult.Option.Validators.Count; j++)
                     {
-                        value += arguments[i].Arity.MaximumNumberOfValues;
+                        optionResult.Option.Validators[j](optionResult);
+                    }
+
+                    if (errorsBefore != SymbolResultTree.ErrorCount)
+                    {
+                        break;
                     }
                 }
 
-                return value;
+                _ = argumentResult.GetArgumentConversionResult();
             }
         }
 
-        internal override bool UseDefaultValueFor(Argument argument) =>
-            FindResultFor(argument) switch
+        private void ValidateArguments(bool completeValidation)
+        {
+            var arguments = Command.Arguments;
+            for (var i = 0; i < arguments.Count; i++)
             {
-                ArgumentResult arg => arg.Argument.HasDefaultValue && 
-                                      arg.Tokens.Count == 0,
-                _ => false
-            };
+                Argument argument = arguments[i];
+
+                if (!completeValidation && !argument.HasDefaultValue)
```

Comment:
It looks like you can just early return from the whole method without entering the loop if `completeValidation == false`.